--- trunk/src/opensips_1_8/modules/snmpstats/alarm_checks.h	2013-04-23 17:33:50.044938309 +0100
+++ alarm_checks.h	2013-04-23 17:47:59.168437507 +0100
@@ -32,7 +32,7 @@
 #ifndef _SNMPSTATS_ALARM_AGENT_
 #define _SNMPSTATS_ALARM_AGENT_
 
-#define ALARM_AGENT_FREQUENCY_IN_SECONDS 5
+#define ALARM_AGENT_FREQUENCY_IN_SECONDS 5 //Jorge. Obsolete, now this parameter is passed in the configuration file
 #define ALARM_AGENT_NAME                 "snmpstats_alarm_agent"
 
 /* Returns the number of bytes currently waiting in the msg queue if they exceed
