diff -Naur opensips-1.8.3-tls/modules/tm/t_msgbuilder.c tm/t_msgbuilder.c
--- opensips-1.8.3-tls/modules/tm/t_msgbuilder.c	2012-05-10 10:21:45.000000000 +0100
+++ tm/t_msgbuilder.c	2013-09-24 11:07:00.595887152 +0100
@@ -181,8 +181,10 @@
 		*len += user_agent_header.len + CRLF_LEN;
 	}
 	/* Content Length, MaxFwd, EoM */
-	*len+=LOCAL_MAXFWD_HEADER_LEN + CONTENT_LENGTH_LEN+1 + (extra?extra->len:0)
-		+ (Trans->extra_hdrs.s?Trans->extra_hdrs.len:0) + CRLF_LEN + CRLF_LEN;
+//	*len+=LOCAL_MAXFWD_HEADER_LEN + CONTENT_LENGTH_LEN+1 + (extra?extra->len:0)
+//		+ (Trans->extra_hdrs.s?Trans->extra_hdrs.len:0) + CRLF_LEN + CRLF_LEN;
+	*len+=CONTENT_LENGTH_LEN+1 + (extra?extra->len:0)
+                + (Trans->extra_hdrs.s?Trans->extra_hdrs.len:0) + CRLF_LEN + CRLF_LEN;
 
 	cancel_buf=shm_malloc( *len+1 );
 	if (!cancel_buf)
@@ -208,8 +210,10 @@
 	append_string( p, cseq_n.s, cseq_n.len );
 	*(p++) = ' ';
 	append_string( p, method->s, method->len );
-	append_string( p, CRLF LOCAL_MAXFWD_HEADER,
-		CRLF_LEN+LOCAL_MAXFWD_HEADER_LEN );
+	//append_string( p, CRLF LOCAL_MAXFWD_HEADER,
+	//	CRLF_LEN+LOCAL_MAXFWD_HEADER_LEN );
+	append_string( p, CRLF,
+		CRLF_LEN );
 
 	/* add Route hdrs (if any) */
 	for ( hdr=buf_hdrs ; hdr ; hdr=hdr->next )
@@ -492,6 +496,7 @@
 	/*headers*/
 	*len += Trans->from.len + Trans->callid.len + to->len +
 		Trans->cseq_n.len + 1 + ACK_LEN + CRLF_LEN + LOCAL_MAXFWD_HEADER_LEN;
+//	*len += Trans->from.len + Trans->callid.len + to->len + Trans->cseq_n.len + 1 + ACK_LEN + CRLF_LEN;
 
 	/* copy'n'paste Route headers */
 	*len += calc_routeset_len(list, cont);
@@ -526,6 +531,8 @@
 	*(p++) = ' ';
 	append_string( p, ACK CRLF LOCAL_MAXFWD_HEADER,
 		ACK_LEN+CRLF_LEN+LOCAL_MAXFWD_HEADER_LEN );
+//	append_string( p, ACK CRLF,
+//		ACK_LEN+CRLF_LEN );
 
 	/* Routeset */
 	p = print_rs(p, list, cont);
@@ -814,8 +821,8 @@
 	*len += CSEQ_LEN + cseq.len + 1 + method->len + CRLF_LEN;
 	/* Route set */
 	*len += calculate_routeset_length(dialog);
-	/* Max-Forwards */
-	*len += LOCAL_MAXFWD_HEADER_LEN;
+	///* Max-Forwards */
+//	*len += LOCAL_MAXFWD_HEADER_LEN;
 	/* Content-Length */
 	*len += CONTENT_LENGTH_LEN + content_length.len + CRLF_LEN;
 	/* Signature */
@@ -844,7 +851,7 @@
 	w = print_routeset(w, dialog);                        /* Route set */
 
 	/* Max-Forwards */
-	append_string(w, LOCAL_MAXFWD_HEADER, LOCAL_MAXFWD_HEADER_LEN);
+//	append_string(w, LOCAL_MAXFWD_HEADER, LOCAL_MAXFWD_HEADER_LEN);
 
 	/* Content-Length */
 	append_string(w, CONTENT_LENGTH, CONTENT_LENGTH_LEN);
