--- trunk/src/opensips_1_8/modules/b2b_entities/dlg.h	2013-03-06 15:15:29.848687838 +0000
+++ b2b_entities/dlg.h	2013-03-06 18:19:50.564437788 +0000
@@ -49,6 +49,7 @@
 
 
 enum b2b_entity_type {B2B_SERVER=0, B2B_CLIENT, B2B_NONE};
+enum fax_error_reply {FAX_NONE=0, FAX_ERR_FOUND=1, FAX_ERR_SENDED=2}; //Added by Jorge
 
 typedef struct b2b_dlginfo
 {
@@ -123,6 +124,7 @@
 	struct socket_info*  send_sock;
 	unsigned int         last_reply_code;
 	int                  db_flag;
+	int                  fax_status;//Added by Jorge; check if b2b receives a request with fax attributes
 }b2b_dlg_t;
 
 typedef struct client_info
