--- trunk/src/opensips_1_8/modules/b2b_logic/b2b_logic.c	2013-03-06 15:14:26.936687769 +0000
+++ b2b_logic/b2b_logic.c	2013-03-06 18:24:13.352687757 +0000
@@ -193,6 +193,12 @@
 
 	LM_DBG("start\n");
 
+   /* load xcoder_b2b_entities api */ //Added by JORGE
+   if(load_xcoder_api(&b2b_xcoder)< 0) {
+      LM_ERR("can't load b2b_xcoder functions\n");
+      return -1;
+   }
+
 	/* load b2b_entities api */
 	if(load_b2b_api(&b2b_api)< 0)
 	{
